### DISCUSSION TITLE
Why did you choose run_sync method name instead of run_async?

### DISCUSSION THREAD

**Discussion by heckad** [sqlalchemy/sqlalchemy#5914] [Ideas]
In sqlalchemy 1.4 introduced the new method [run_sync](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.run_sync). When combined with `await`, it reads very strangely. Maybe rename to `run_async`. This is more clearer. We run something asynchronously.

**Comment by heckad** [comment]
Also for this method not needed add args and kwargs because for bound we can use lambda. For example 
```
await conn.run_async(lambda: metadata.create_all())
```

**Comment by CaselIT** [comment]
Hi,

> We run something asynchronously.

Actually the function inside `run_sync` is run with a synchronous style api, so without using await etc. That is what gives the name to the function

**Comment by heckad** [comment]
Hi, 

> Actually the function inside `run_sync` is run with a synchronous style api, so without using await etc. That is what gives the name to the function

But we, as users, look at it from the user's point of view. From the user's point of view, we take a callback and launch it not now, but sometime, but now we just get a future like object. And how it's done inside is already implementation details.

**Comment by zzzeek** [comment]
if the function were named after its implementation detail it would be called `run_in_greenlet()`.

a more correct name would be `run_synchronous_blocking_style_code_in_asyncio()` but that's kind of long.   `run_async()` definitely is misleading because you're already in asyncio, everything is "async".   saying "await" already means you are running something asynchronously.

**Comment by CaselIT** [comment]
> 
> 
> Hi,
> 
> > Actually the function inside `run_sync` is run with a synchronous style api, so without using await etc. That is what gives the name to the function
> 
> But we, as users, look at it from the user's point of view. From the user's point of view, we take a callback and launch it not now, but sometime, but now we just get a future like object. And how it's done inside is already implementation details.

I kind of see your point. Sill I don't think `run_async` would not be correct. Maybe a `run_sync_function` would better convey what the call does

**Comment by heckad** [comment]
> `run_async()` definitely is misleading because you're already in asyncio,

`run_async` has nothing to do with `asyncio`. It's like a pattern. There is a run method that runs the code synchronously (for example, in a virtual machine) and a `run_async` method that runs the code asynchronously (most often it is used with a thread pool).

> `run_synchronous_blocking_style_code_in_asyncio()`

Why is it necessary `synchronous_blocking_style_code`? You can call with lambda wich do something like `loop.run_untill_compleate(some_coro)`. You just call any code in sometimes future and just get an object which provides methods for manipulating this task.

**Comment by heckad** [comment]
In our projects we use 
```
"""
The Background Common module.

Defines primitives for executing I/O blocking operations in the background using
a thread pool executor.
"""
import typing
from typing import Callable

from config import settings

if typing.TYPE_CHECKING:
    R = typing.TypeVar('R')

import asyncio
import time
from concurrent.futures import ThreadPoolExecutor

from loguru import logger

__all__ = [
    'run_async',
]

###
# Thread pool executor used strictly to submit queries/operations to the database
# in background.
#
DB_THREAD_POOL_EXECUTOR = ThreadPoolExecutor(
    max_workers=int(settings.DB['thread_pool_executor_size']),
    thread_name_prefix='db_worker_',
)


async def run_async(func: "Callable[[], R]") -> "R":
    """
    Runs a callable on the database thread pool executor using the current thread's
      I/O loop instance.

    Usage:
        def blocking_task(arg1, arg2, arg3):
            # ...
            # does some blocking stuff
            # ...

        run_async(lambda: blocking_task(arg1, arg2, arg3))

    :param Callable[[], R] func: the function or callable to run in background
    :return: func's return value
    """
    loop = asyncio.get_event_loop()

    def work():
        start = time.monotonic()
        result = func()
        end = time.monotonic()

        if end - start > 5:
            duration = '%.2f' % (end - start)
            logger.warning(
                f'slow_query_logger: {duration}s for executing {func}'
            )
        return result

    return await loop.run_in_executor(DB_THREAD_POOL_EXECUTOR, work)
```

This code was borrowed [from](https://github.com/aalhour/cookiecutter-aiohttp-sqlalchemy) this repository.
.

**Comment by zzzeek** [comment]
> > `run_async()` definitely is misleading because you're already in asyncio,
> 
> `run_async` has nothing to do with `asyncio`. It's like a pattern. There is a run method that runs the code synchronously (for example, in a virtual machine) and a `run_async` method that runs the code asynchronously (most often it is used with a thread pool).

that's not what run_sync() does here.

> 
> > `run_synchronous_blocking_style_code_in_asyncio()`
> 
> Why is it necessary `synchronous_blocking_style_code`? You can call with lambda wich do something like `loop.run_untill_compleate(some_coro)`. 

run_sync() does not accept a coroutine, is accepts a plain Python function which calls blocking style API functions that ultimately call awaitables for the actual IO.     it's important to note the difference in terminology between "blocking **style**" and "**blocking**".   SQLAlchemy's asyncio extension does not run "blocking" IO calls.   The `run_sync()` method runs code that is written in "blocking **style**" but ultimately is using asyncio using greenlets.


> You just call any code in sometimes future and just get an object which provides methods for manipulating this task. 

if you call upon SQLAlchemy's async-adapted database drivers without using await you get an error message.

Just so we're on the same page, make sure you try out an example script like https://github.com/sqlalchemy/sqlalchemy/blob/master/examples/asyncio/basic.py .  

For example, your proposal "loop.run_until_complete(lambda)" is not compatible because the event loop is already running:

```diff
diff --git a/examples/asyncio/basic.py b/examples/asyncio/basic.py
index efdb7e9e8..c7576f52e 100644
--- a/examples/asyncio/basic.py
+++ b/examples/asyncio/basic.py
@@ -39,8 +39,9 @@ async def async_main():
         # version of the AsyncConnection object to any synchronous method,
         # where synchronous IO calls will be transparently translated for
         # await.
-        await conn.run_sync(meta.drop_all)
-        await conn.run_sync(meta.create_all)
+
+        asyncio.get_event_loop().run_until_complete(lambda: meta.drop_all(conn.sync_connection))
+        asyncio.get_event_loop().run_until_complete(lambda: meta.create_all(conn.sync_connection))
 
         # for normal statement execution, a traditional "await execute()"
         # pattern is used.

```

output:

    RuntimeError: This event loop is already running

> In our projects we use

it looks like you are running a plain ThreadPoolExecutor to run blocking database code.   That is not at all what run_sync() does.   run_sync() doesn't run blocking IO and it does not need to use executors or anything with thread-pools.

**Comment by heckad** [comment]
Yes, you are right. I have not seen how it is implemented.

In my example, I mean something like 
```
await conn.run_sync(lambda: asyncio.run(some_another_coro()))
```
It would have worked if used ThreadPool instead of greenlet.

**Comment by zzzeek** [comment]
OK, anyway, our API is not given a cotourine it's adapting a non-coroutine, but one that ultimately uses *internal* coroutines for IO, into an awaitable, and then invoking it.    With a few obscure exceptions, there is no comparison to how this works to nearly anything else in the Python world right now.    A name that might be more descriptive without being too long might be `run_as_awaitable()`.      The way the function runs is most similar to how one would see code running under gevent or eventlet, except it calls into a pure asyncio IO library.

**Comment by CaselIT** [comment]
@zzzeek maybe we should make clear that `run_sync` does not run inside a thread executor and that doing any other blocking call that is not sqlalchemy, like sleep/requests, will block the asyncio loop

**Comment by zzzeek** [comment]
the method is described in great detail here: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=run_sync#adapting-orm-lazy-loads-to-asyncio    so the docstring should at least link to that

**Comment by CaselIT** [comment]
> Also for this method not needed add args and kwargs because for bound we can use lambda. 

Since the lambda can be used either way, I don't see what's the harm in having args and kwargs.

> 
> 
> the method is described in great detail here: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=run_sync#adapting-orm-lazy-loads-to-asyncio so the docstring should at least link to that

Will update the docs.

Also I think we could transform this to a discussion

**Comment by CaselIT** [comment]
> Will update the docs.

https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2453

**Comment by zzzeek** [comment]
just FTR we should probably change the name to something, I was not that happy with run_sync() anyway.    run_as_awaitable() seems closest to me for now and at least wont get confused with the ambiguity of the terms "sync/ async".

**Comment by heckad** [comment]
Maybe make run_sync as the internal method? We can check we are in async code or not using
```
import asyncio


def check_we_are_in_async_code():
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return False
    else:
        return True


print("Outside async code:", check_we_are_in_async_code())


async def main():
    print("Inside async code:", check_we_are_in_async_code())


if __name__ == '__main__':
    asyncio.run(main())
``` 

Let's make decorator wich do something like
```
def sync_to_async(func):
    @wraps(func)
    def decorated(*args, **kwargs):
        if check_we_are_in_async_code():
            return run_sync(func, *args, **kwargs)
        else:
            return func(*args, **kwargs)

    return decorated
```
and decorate all methods which should work as async.

**Comment by heckad** [comment]
Workable example:
```
import asyncio
from _contextvars import ContextVar
from functools import partial
from functools import wraps

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.ext.asyncio import create_async_engine

current_connection = ContextVar('current_connection')


def check_we_are_in_async_code():
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return False
    else:
        return True


def sync_to_async_method_with_bind(func):
    @wraps(func)
    def decorated(self, *args, **kwargs):
        if check_we_are_in_async_code():
            conn = current_connection.get()
            return conn.run_sync(partial(func, self), *args, **kwargs)
        else:
            return func(self, *args, **kwargs)

    return decorated


MetaData.drop_all = sync_to_async_method_with_bind(MetaData.drop_all)
meta = MetaData()

t1 = Table(
    "t1", meta, Column("id", Integer, primary_key=True), Column("name", String)
)


async def async_main():
    # engine is an instance of AsyncEngine
    engine = create_async_engine(
        'sqlite:///:memory:',
        echo=True,
    )

    # conn is an instance of AsyncConnection
    async with engine.begin() as conn:
        token = current_connection.set(conn)  # Do this in __aenter__ method

        await meta.drop_all()

        current_connection.reset(token) # Do this in __aexit__ method


asyncio.run(async_main())
```

Also, we can choose by connection class using `isinstance(conn, AsyncConnection)` for check in the decorator.

**Comment by heckad** [comment]
Universal code for async and sync mode:
```
import asyncio
from _contextvars import ContextVar
from functools import partial
from functools import wraps

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.ext.asyncio import AsyncConnection
from sqlalchemy.ext.asyncio import create_async_engine

current_connection = ContextVar('current_connection')


def sync_to_async_method_with_bind(func):
    @wraps(func)
    def decorated(self, *args, **kwargs):
        conn = current_connection.get()

        if isinstance(conn, AsyncConnection):
            return conn.run_sync(partial(func, self), *args, **kwargs)
        else:
            return func(self, conn, *args, **kwargs)

    return decorated


MetaData.drop_all = sync_to_async_method_with_bind(MetaData.drop_all)
meta = MetaData()

t1 = Table(
    "t1", meta, Column("id", Integer, primary_key=True), Column("name", String)
)


async def async_main():
    # engine is an instance of AsyncEngine
    engine = create_async_engine(
        'sqlite:///:memory:',
        echo=True,
    )

    # conn is an instance of AsyncConnection
    async with engine.begin() as conn:
        token = current_connection.set(conn)  # Do this in __aenter__ method

        await meta.drop_all()

        current_connection.reset(token)  # Do this in __aexit__ method


asyncio.run(async_main())


def main():
    # engine is an instance of Engine
    engine = create_engine(
        'sqlite:///:memory:',
        echo=True,
    )

    # conn is an instance of AsyncConnection
    with engine.begin() as conn:
        token = current_connection.set(conn)  # Do this in __aenter__ method

        meta.drop_all()

        current_connection.reset(token)  # Do this in __aexit__ method

main()
```

**Comment by CaselIT** [comment]
I'm not sure what you what to achieve here. Could you explain?

Also note that sqlite cannot be used with the asyncio extension.

**Comment by heckad** [comment]
I want to get rid of the `run_sync` method. If we use async engine context then all blocking style methods should make available interior without user help because a user can forget use run_sync method. One solution is to use context variables that can work as both thread-local and task-local and chose what return by this connection type. If a connection type is async we return something that returns `run_sync` method another we do old behaviour.

**Comment by zzzeek** [comment]
> I want to get rid of the `run_sync` method. If we use async engine context then all blocking style methods should make available interior without user help because a user can forget use run_sync method.

OK, this seems to be a different issue than what you originally raised, so now you're saying that when one uses the SQLAlchemy async API , that we have in our examples some utility methods that aren't available as async directly so we provide an adapter, you would rather not have the adapter be present, you would rather have all IO related methods throughout all of SQLAlchemy be available with an "await" API.      

It's not clear if you're proposing that SQLAlchemy make this available by deafult, e.g. metadata.create_all() would work "out of the box" inside of an asyncio script with "await", or if you are proposing that the end user be given a special decorator that they would apply to methods such as metadata.create_all, that they would then monkeypatch onto metadata, so that within their code they can say "await metadata.create_all()" rather than "await conn.run_sync(metadata.create_all)".

The first option, if applied, would not have to use context variables or anything else, we would simply provide an AsyncMetaData object that has awaitable "create_all()" and "drop_all()" methods.      

This option is not feasible because each of SQLAlchemy's async APIs are presented as a new module with classes that wrap around existing ones, and there's simply no need for this to always have to extend to DDL-related methods and other assorted methods throughout the library, that places an undue burden for development, testing, documentation, and maintenance.     Additionally, the event system is not adapted to async at all, so if one were using events like before_execute() or before_cursor_execute() , as well as ORM events like before_insert(), after_insert(), etc., the connection passed within those is a sync connection.  So there is no solution to the issue of SQLAlchemy providing a complete "asyncio for absolutely everything everywhere" approach, this would be impractical and is unnecessary for the stated goals of Python's asyncio library.  SQLAlchemy's asyncio API provides the asyncio interface for the part that it's relevant, e.g. within the DML and DQL emitting portions of the Core and ORM components.   people don't need to write SQL migration tools and routines using asyncio.

The second option is not feasible, as an addition to SQLAlchemy itself (you are of course free to use any pattern you'd like in your own application) for reasons which include the following:   1. we would never encourage monkeypatching 2. it's unnecessarily complex, as a much simpler approach already exists  3. it associates connections with implicit context variables which is the opposite of SQLAlchemy's design goals.   connections and sessions are always explicitly passed and global "contexts" are typically a last resort programming pattern IMO.

> should make available interior without user help because a user can forget use run_sync method.

users would not forget to use run_sync because the code raises an error for a variety of reasons. 

1.   create_all() and drop_all() require a connection to be passed, because the metatdata is not bound.   so that raises this error:
    
    sqlalchemy.exc.UnboundExecutionError: MetaData object is not bound to an Engine or Connection.  Execution can not proceed without a database to execute against.


2. try passing the async connection to them, as `meta.drop_all(conn)`.  That raises an error, because AsyncConnection is not accepted by these (this error message should be improved):
   
    AttributeError: 'AsyncConnection' object has no attribute '_run_ddl_visitor'

3. Try passing the sync connection instead, that is, `meta.drop_all(conn.sync_connection)`.   the greenlet adaption will fail:

    sqlalchemy.exc.InvalidRequestError: greenlet_spawn has not been called; can't call await_() here.

Not to mention if the method were to silently fail, their code would be failing anyway.   So there is no chance that forgetting to call run_sync() in the appropriate place would not raise an error and the user would have a clear path to fixing their code.    Python's asyncio API is already fraught with a significant number of ways to forget to do things correctly in any case, it is already a confusing API and I am confident that users will have no problem using run_sync() ( or wahtever we call it) for the extremely small number of cases where it's needed.

To sum up I can easily agree that the name "run_sync()" is not clear, so we can change the name, but it is only used for very occasional DDL-related calls and is not needed for nearly anything else.   it provides a clear way to explicitly bridge into any arbitrary non-awaitable code and is useful far beyond the small number of example cases we gave.    it's in fact how the whole asyncio->non asyncio->asyncio bridge works.
